keras/layers/wrappers.py

```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
+                else:
+                    raise Exception("Mask is of an unexpected shape. Mask's ndim: %s, input's ndim %s" %
+                                    (mask_ndim, input_ndim))
+                mask_i = K.reshape(mask_i, (-1,) + mask_shape[2:])  # (batch_size * timesteps, ...)
+            reshaped_xs.append(reshaped_x)
+            reshaped_masks.append(mask_i)
+        if len(inputs) == 1:
+            reshaped_xs = reshaped_xs[0]
+            reshaped_masks = reshaped_masks[0]
+        return reshaped_xs, reshaped_masks
+
     def call(self, inputs, mask=None):
-        input_shape = K.int_shape(inputs)
-        if input_shape[0]:
-            # batch size matters, use rnn-based implementation
-            def step(x, _):
-                output = self.layer.call(x)
+        if not isinstance(inputs, list):
+            inputs = [inputs]
+            mask = [mask]
+        timesteps = K.int_shape(inputs[0])[1]
+        input_shape = [K.int_shape(input_i) for input_i in inputs]
+        if len(inputs) == 1:
+            input_shape = input_shape[0]
+        first_input_shape = self.input_spec[0].shape
+        if len(inputs) == 1 and first_input_shape[0]:
+            # The batch size is passed when defining the layer in some cases (for
+            # example if it is stateful).  We respect the input shape in that
+            # case and don't reshape the input. This is slower.  K.rnn also
+            # expects only a single tensor, so we can't do this if we have
+            # multiple inputs.
+            def step(input_i, states):
+                output = self.layer.call(input_i)
                 return output, []
-
-            _, outputs, _ = K.rnn(step, inputs,
-                                  initial_states=[],
-                                  input_length=input_shape[1],
-                                  unroll=False)
-            y = outputs
+            _, outputs, _ = K.rnn(step, inputs, mask=mask, input_states=[])
         else:
-            # No batch size specified, therefore the layer will be able
-            # to process batches of any size.
-            # We can go with reshape-based implementation for performance.
-            input_length = input_shape[1]
-            if not input_length:
-                input_length = K.shape(inputs)[1]
-            # Shape: (num_samples * timesteps, ...)
-            inputs = K.reshape(inputs, (-1,) + input_shape[2:])
-            y = self.layer.call(inputs)  # (num_samples * timesteps, ...)
-            # Shape: (num_samples, timesteps, ...)
+            reshaped_xs, reshaped_masks = self.reshape_inputs_and_masks(inputs, mask)
+            outputs = self.layer.call(reshaped_xs, mask=reshaped_masks)
             output_shape = self.compute_output_shape(input_shape)
-            y = K.reshape(y, (-1, input_length) + output_shape[2:])
-
-        # Apply activity regularizer if any:
-        if (hasattr(self.layer, 'activity_regularizer') and
```

Comment:
Looks like this was added after I branched off from keras with this layer.  I'll add these lines back in.